xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_dof.cpp

```diff
@@ -0,0 +1,130 @@
+#include "stdafx.h"
+#include "rasterizer_dx9_dof.h"
+
+#include "rasterizer_dx9_main.h"
+#include "rasterizer_dx9_submit.h"
+
+/* structures */
+
+struct s_convolution_d3d_vertex
+{
+    real32 pos_x;
+    real32 pos_y;
+    real32 pos_z;
+    real32 bounds_x;
+    real32 bounds_y;
+    D3DCOLOR color;
+};
+
+/* prototypes */
+
+e_rasterizer_target __cdecl sub_87E570(real32 unused1, real32 a1, real32 unused2, e_rasterizer_target a2, e_rasterizer_target a3, bool a4);
+
+/* public code */
+
+void __cdecl rasterizer_dx9_dof_render_bloom(real32 blur_amount, real32 threshold, real32 brightness, real32 box_factor, real32 max_factor, bool a6, bool a7)
+{
+	INVOKE(0x26C3EB, 0x0, rasterizer_dx9_dof_render_bloom, blur_amount, threshold, brightness, box_factor, max_factor, a6, a7);
+	return;
+}
+
+void __cdecl rasterizer_dx9_render_depth_of_field(real32 lower, real32 upper, real32 a3)
+{
+	INVOKE(0x26CABC, 0x0, rasterizer_dx9_render_depth_of_field, lower, upper, a3);
+	return;
+}
+
+e_rasterizer_target rasterizer_dx9_convolve_screen_surfaces(
+    real32 a1,
+    real32 a2,
+    real32 a3,
+    e_rasterizer_target source,
+    e_rasterizer_target destination,
+    e_rasterizer_target stretch_source,
+    e_rasterizer_target stretch_destination,
+    int32 a8,
+    int32 a9,
+    real32 a10,
+    real32 a11)
+{
+    e_rasterizer_target source_target = source;
+    e_rasterizer_target stretch = (stretch_source == _rasterizer_target_none ? source : stretch_source);
+    real32 acc = 0.f;
+
+    if (a1 > 0.f)
+    {
+        e_rasterizer_target dst = destination;
+        do
+        {
+            real32 res = a1 - acc;
+            res = PIN(res, 0.f, 1.f);
+
+            if (res > 0.f)
+            {
+                source_target = sub_87E570(res, a2, a3, source_target, dst, a8);
+                
+                // Swap temp and stretch values
+                e_rasterizer_target temp = stretch;
+                stretch = dst;
+                dst = temp;
+            }
+            a2 *= (1.f - a3 + a3 + a3);
+            acc += 1.f;
+        }
+        while (acc < a1);
+    }
+
+    if (stretch_destination != source_target && stretch_destination != _rasterizer_target_none)
+    {
+        rasterizer_dx9_copy_target(source_target, stretch_destination);
+        source_target = stretch_destination;
+    }
+
+    return source_target;
+
+    /*
+    return INVOKE(0x26C2CA, 0x0, rasterizer_dx9_convolve_screen_surfaces,
+        a1,
+        a2,
+        a3,
+        source,
+        destination,
+        stretch_source,
+        stretch_destination,
+        a8,
+        a9,
+        a10,
+        a11);
+    */
+}
+
+void rasterizer_dx9_draw_convolution_rect(uint32 screen_resolution_x, uint32 screen_resolution_y, D3DCOLOR color)
```

Comment:
this just draws a single color quad, has nothing to do with convolution or anything in particular with convolution

Reply:
rasterizer_dx9_draw_overlay_rect?